feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksScreen.kt

```diff
@@ -79,41 +109,60 @@ fun BookmarksScreen(
+        )
+        item(span = { GridItemSpan(maxLineSpan) }) {
+            Spacer(Modifier.windowInsetsBottomHeight(WindowInsets.safeDrawing))
+        }
+    }
+}
 
-        when (feedState) {
-            is NewsFeedUiState.Loading -> {
-                item(span = { GridItemSpan(maxLineSpan) }) {
-                    NiaLoadingWheel(
-                        modifier = Modifier
-                            .fillMaxWidth()
-                            .wrapContentSize()
-                            .testTag("forYou:loading"),
-                        contentDesc = stringResource(id = R.string.saved_loading),
-                    )
-                }
-            }
+@Composable
+private fun EmptyState(modifier: Modifier = Modifier) {
+    Column(
+        modifier = modifier
+            .padding(16.dp)
+            .fillMaxSize()
+            .testTag("bookmarks:empty"),
+        verticalArrangement = Arrangement.Center,
+        horizontalAlignment = Alignment.CenterHorizontally
+    ) {
+        Image(
+            modifier = Modifier.fillMaxWidth(),
+            painter = painterResource(id = R.drawable.img_empty_bookmarks),
+            contentDescription = null
+        )
 
-            is NewsFeedUiState.Success -> {
-                if (feedState.feed.isNotEmpty()) {
-                    newsFeed(
-                        feedState = feedState,
-                        onNewsResourcesCheckedChanged = { id, _ -> removeFromBookmarks(id) },
-                    )
-                    item(span = { GridItemSpan(maxLineSpan) }) {
-                        Spacer(Modifier.windowInsetsBottomHeight(WindowInsets.safeDrawing))
-                    }
-                } else item(span = { GridItemSpan(maxLineSpan) }) {
-                    Box(
-                        modifier = Modifier
-                            .fillMaxHeight()
-                            .wrapContentSize()
-                            .testTag("bookmarks:empty"),
-                        contentAlignment = Alignment.Center
-                    ) {
-                        Text(text = stringResource(id = R.string.bookmarks_empty))
-                    }
-                }
-            }
-        }
+        Spacer(modifier = Modifier.height(16.dp))
+
+        Text(
+            text = stringResource(id = R.string.bookmarks_empty_error),
+            modifier = Modifier.fillMaxWidth(),
+            textAlign = TextAlign.Center,
+            style = MaterialTheme.typography.titleMedium,
+            fontWeight = FontWeight.Bold
+        )
+
+        Spacer(modifier = Modifier.height(5.dp))
+
+        Text(
+            text = stringResource(id = R.string.bookmarks_empty_description),
+            modifier = Modifier.fillMaxWidth(),
+            textAlign = TextAlign.Center,
+            style = MaterialTheme.typography.bodyMedium
+        )
     }
 }
+
+@Composable
+private fun LoadingState(modifier: Modifier = Modifier) {
```

Comment:
Amazing separation :) just one thing I always care about is priorities. For instance, the `when` expression checks for the `Loading` state **first**, and you put the `Loading` state composable function at the end of the file. I wouldn't do that, I would put the `Loading()` function before the `BookmarksGrid()`.

- This becomes very handy when we have a huge class (or we will have?) and we are trying to scroll down and find the right function for it, so I am always placing those functions based on the priorities and the places where they are getting called.

Please let me know what you think :) thank you.

Reply:
Great point, I changed the order and added some previews as well :)